File: Trees_Graphs/go/list_of_depths.go
package main
import (
  "types"
)


func writeDepth(node *types.TreeNode, dNum int, list *[]*types.LinkedListOfTreeNodes) {
  var depthList types.LinkedListOfTreeNodes
  if dNum == len(*list) {
    depthList.Append(node)
    *list = append(*list, &depthList)
  } else {
    depthList = *(*list)[dNum]
    depthList.Append(node)
    (*list)[dNum] = &depthList
  }

  if node.LeftChild != nil {
    writeDepth(node.LeftChild, dNum + 1, list)
  }

  if node.RightChild != nil {
    writeDepth(node.RightChild, dNum + 1, list)
  }
}


func main() {
  array := []int{1, 2, 3, 4, 5, 6, 7, 8}
  var depths []*types.LinkedListOfTreeNodes
  var depthList types.LinkedListOfTreeNodes

  root := types.CreateTree(array)
  depthList.Append(root)
  depths = append(depths, &depthList)

  writeDepth(root.LeftChild, 1, &depths)
  writeDepth(root.RightChild, 1, &depths)

  for i := 0; i < len(depths); i++ {
    types.PrintLinkedListOfTreeNodes(*depths[i])
  }
}

File: Trees_Graphs/python/list_of_depths.py
from Tree import *
from LinkedList import *


def write_depth(node, depth_num, depths):
    depth_list = LinkedList()

    if depth_num == len(depths):
        depth_list.append(node)
        depths.append(depth_list)
    else:
        depth_list = depths[depth_num]
        depth_list.append(node)
        depths[depth_num] = depth_list

    if node.left_child:
        write_depth(node.left_child, depth_num + 1, depths)

    if node.right_child:
        write_depth(node.right_child, depth_num + 1, depths)


def main():
    depths = []
    depth_list = LinkedList()
    array = list(range(7))
    tree = Tree(array)

    depth_list.append(tree.root)
    depths.append(depth_list)

    write_depth(tree.root.left_child, 1, depths)
    write_depth(tree.root.right_child, 1, depths)

    for depth in depths:
        print(depth)


if __name__ == '__main__':
    main()
